intermediate_source/rpc_param_server_tutorial.rst

```diff
@@ -0,0 +1,398 @@
+               out = self.model(inp)
+               # This output is forwarded over RPC, which as of 1.5.0 only accepts CPU tensors.
+               # Tensors must be moved in and out of GPU memory due to this.
+               out = out.to("cpu")
+               return out
+
+Next, we'll define a few miscellanous functions useful for training and verification purposes. The first, ``get_dist_gradients``\ , will take in a Distributed Autograd context ID and call into the ``dist_autograd.get_gradients`` API in order to retrieve gradients computed by distributed autograd. More information can be found in the `distributed autograd documentation <https://pytorch.org/docs/stable/rpc.html#distributed-autograd-framework>`_. Note that we also iterate through the resulting dictionary and convert each tensor to a CPU tensor, as the framework currently only supports sending tensors over RPC. Next, ``get_param_rrefs`` will iterate through our model parameters and wrap them as a (local) `RRef <https://pytorch.org/docs/stable/rpc.html#torch.distributed.rpc.RRef>`_. This method will be invoked over RPC by trainer nodes and will return a list of the parameters to be optimized on this node. This is required as input to the `Distributed Optimizer <https://pytorch.org/docs/stable/rpc.html#module-torch.distributed.optim>`_\ , which requires all parameters it must optimize as a list of ``RRef``\ s. 
+
+.. code-block:: python
+
+   # Use dist autograd to retrieve gradients accumulated for this model.
+       # Primarily used for verification.
+       def get_dist_gradients(self, cid):
+           grads = dist_autograd.get_gradients(cid)
+           # This output is forwarded over RPC, which as of 1.5.0 only accepts CPU tensors.
+           # Tensors must be moved in and out of GPU memory due to this.
+           cpu_grads = {}
+           for k, v in grads.items():
+               k_cpu, v_cpu = k.to("cpu"), v.to("cpu")
+               cpu_grads[k_cpu] = v_cpu
+           return cpu_grads
+
+       # Wrap local parameters in a RRef. Needed for building the
+       # DistributedOptimizer which optimizes paramters remotely.
+       def get_param_rrefs(self):
+           param_rrefs = [rpc.RRef(param) for param in self.model.parameters()]
+           return param_rrefs
+
+Finally, we'll create methods to initialize our parameter server. Note that there will only be one instance of a parameter server across all processes, and all trainers will talk to the same parameter server and update the same stored model. As seen in ``run_parameter_server``\ , the server itself does not take any independent actions; it waits for requests from trainers (which are yet to be defined) and responds to them by running the requested function.
+
+.. code-block:: python
+
+   # The global parameter server instance.
+   param_server = None
+   # A lock to ensure we only have one parameter server.
+   global_lock = Lock()
+
+
+   def get_parameter_server(num_gpus=0):
+     """
```

Comment:
mismatch indent

Reply:
Fixed, the rest of the code was also overindented.